### DISCUSSION TITLE
Serving custom files

### DISCUSSION THREAD

**Discussion by florian-lefebvre** [nuxt/nuxt#17807] [Questions] [answered]
I wrote a custom sitemap module (https://github.com/florian-lefebvre/portfolio/blob/v4_nuxtjs/modules/sitemap.ts) to generate a `sitemap.xml` file under `.output/public` after building the app in production using `yarn build`. To do that, I use the `build:done` hook.
The file is created with the correct content where it should but it's not served at `http://localhost:3000/sitemap.xml` when doing `yarn start`.
What can I do?

**Comment by florian-lefebvre** [comment]
I found a solution: I added an empty `public/sitemap.xml` so it gets overwritten on build but taken into account by Nuxt.

**Reply by danielroe** [reply]
That's not ideal as Nitro calculates a hash based on the file content, and although it will serve the file it won't serve the correct headers.

I'll look at the hook ordering, but I'd try using `nitro: generate` instead. We're also likely to expose a way from nitropack for you to add static files directly to your build.

**Reply by florian-lefebvre** [reply]
I just tried to update the hook from `build:done` to `nitro:generate` but I get an error:

```bash
√ Server built in 2404ms                                                                                                                                  20:07:57
i Nitro preset is server                                                                                                                                  20:07:57
i Cleaning up .output                                                                                                                                     20:07:57  
start Generating public...                                                                                                                                20:07:57

 ERROR  ENOENT: no such file or directory, open 'D:\DEV\PROJECTS\PERSO\florian-lefebvre.dev\portfolio\.output\public\sitemap.xml'                         20:07:57  

error Command failed with exit code 1.
```

I think the issue is with the resolver but I don't know how to fix it exactly.

```ts
const resolver = createResolver(import.meta.url);
const filePath = await resolver.resolvePath(
  "~/.output/public/sitemap.xml"
);
```

**Reply by florian-lefebvre** [reply]
In fact it doesn't work on Vercel, it throws the same error than in my previous comment

**Reply by danielroe** [reply]
Yes, `resolvePath` won't work as there is no file there before you copy it across.

Here's an example of how you might do it: https://github.com/nuxt-community/partytown-module/blob/main/src/module.ts#L99-L102

**Reply by florian-lefebvre** [reply]
Thanks, I'll have a look

**Reply by florian-lefebvre** [reply]
Hi @danielroe!
So I now have something working:
```ts
import { createResolver, defineNuxtModule } from "@nuxt/kit";
import { promises as fsp } from "fs";
import { NuxtPage } from "~/node_modules/@nuxt/kit/node_modules/@nuxt/schema/dist/index";
import { SitemapStream, streamToPromise } from "sitemap";
import { Readable } from "stream";
import consola from "consola";
import { join, dirname } from "path";
import serveStatic from "serve-static";

interface Options {
  hostname: string;
  dynamicRoutes: Promise<string[]> | string[];
}

const DEFAULTS: Options = {
  hostname: "http://localhost:3000",
  dynamicRoutes: [],
};

export default defineNuxtModule({
  meta: {
    name: "sitemap",
    version: "0.0.1",
    configKey: "sitemap",
    compatibility: {
      bridge: false,
    },
  },
  defaults: DEFAULTS,
  async setup(options, nuxt) {
    let routes: NuxtPage[];

    const generateSitemp = async (path: string): Promise<void> => {
      const sitemapRoutes: string[] = [
        ...routes
          .filter((route) => !route.path.includes(":"))
          .map((route) => route.path),
        ...(await options.dynamicRoutes),
      ];

      const sitemapConfig = {
        hostname: options.hostname,
        urls: sitemapRoutes,
      };

      const stream = new SitemapStream(sitemapConfig);
      const sitemap = await streamToPromise(
        Readable.from(sitemapConfig.urls).pipe(stream)
      ).then((data) => data.toString());

      const dirPath = dirname(path);
      await fsp.mkdir(dirPath, { recursive: true });

      await fsp.writeFile(path, sitemap);
    };

    const handleSitemapGeneration = async (fn: Function): Promise<void> => {
      try {
        await fn();
        consola.success(`Sitemap generated`);
      } catch (e) {
        consola.error("Error generating sitemap");
      }
    };

    nuxt.hook("pages:extend", async (pages) => {
      routes = pages;
    });

    if (nuxt.options.dev) {
      nuxt.hook("build:done", async (builder) => {
        await handleSitemapGeneration(async () => {
          const resolver = createResolver(import.meta.url);
          const filePath = await resolver.resolvePath("~/.sitemap/sitemap.xml");
          await generateSitemp(filePath);
          // nuxt.options.serverMiddleware.push({
          //   path: "/",
          //   handler: serveStatic(dirname(filePath)),
          // });
        });
      });
    } else {
      nuxt.hook("nitro:generate", async (ctx) => {
        await handleSitemapGeneration(async () => {
          const filePath = join(ctx.output.publicDir, "sitemap.xml");
          await generateSitemp(filePath);
        });
      });
    }
  },
});
```
However, setting the server middleware this way (like in the partytown module) doesn't work. But if I set it like below, it works:
```ts
serverMiddleware: [
    {
      path: "/",
      handler: serveStatic(__dirname + "/.sitemap"),
    },
  ],
```

Do you know why?

**Reply by danielroe** [reply]
Nitro needs to be compiled, and we need a file/package handler as a source, rather than an inline function. (Inline functions will work, but only in dev mode.) The reason yours doesn't is that adding it to serverMiddleware after the build is done is too late. By that point the nitro app is already up-and-running.

**Reply by florian-lefebvre** [reply]
Thanks it works now! Thanks a lot for taking the time to help me

**Comment by florian-lefebvre** [comment, marked as answer]
Here is the working code that actually solves the problem: [Link to file](https://github.com/florian-lefebvre/portfolio/blob/c513428dea912a19ffb684b8b571b08b8882158c/modules/sitemap.ts)

**Comment by madsh93** [comment]
Well done building a sitemap module. Do you plan to publish it as a package? 🥇

**Reply by florian-lefebvre** [reply]
No, it's temporary while waiting for the official one: https://github.com/nuxt-community/sitemap-module.
But feel free to use my code, I'm going to work on getting all routes automatically (for dynamic ones especially) this afternoon.

**Comment by benoitdemaegdt** [comment]
Hey there 👋 
Thank you @florian-lefebvre for this code. I used it in my own Nuxt3 app and it is now working great 👌 

I wrote the steps one should follow to get it working. Here are the instructions : https://github.com/benoitdemaegdt/nuxt3-sitemap
I hope this can help someone else 😉

**Reply by florian-lefebvre** [reply]
Hi great! (and thx for the credits 😄)
Out of curiosity, why removing the ability to use the module in dev mode?

**Reply by benoitdemaegdt** [reply]
It is indeed a minimal version compared to what you wrote 👌 
There are few reasons behind this : 
- I'm not used to Nuxt Modules development and wanted something as short and simple as possible.
- I do not need a sitemap in dev mode (I only use it for the search console & SEO).

Also, this version does not handle dynamic pages (I think your version does ?). I'll probably do an update when @nuxt/content v2 will be released (hopefully this should happen soon 🤞 ).

All that said, I guess it is probably a good practice for a Nuxt Module to behave in the same way in prod & dev mode.
I added a link to this discussion and to your original file in the READ.me. That should help someone looking for more advanced features 😉

**Reply by benoitdemaegdt** [reply]
hi @florian-lefebvre 
Just to let you know that nuxt3 deprecated `buildModules`
See commit here : https://github.com/nuxt/framework/pull/3830/files

`modules` should be used instead

**Reply by florian-lefebvre** [reply]
Thanks for the advice!
